Rework shard snapshot workers

pxsalehi opened the pull request:
Currently, when starting the shard snapshot tasks, we loop through the
shards to be snapshotted back-to-back. This will calculate whether/which
changes in the shard there are to be snapshotted, write some metadata,
and ONLY if there are changes and there are files to be uploaded, fork
off the file upload calls. This could be further improved by parallelizing 
shard snapshot tasks (currently, only file uploads happen in parallel). 

This change uses a worker pool and a queue to:

1. Paralleize (with limited number of workers) shard snapshotting, and 
    limit the number of the concurrently running snapshot tasks.
2. Prioritize shard snapshot tasks over file snapshot tasks.

Closes #83408

pxsalehi committed:
Fork off shard snapshot calls in parallel

pxsalehi committed:
Update docs/changelog/88209.yaml

elasticmachine commented:
Pinging @elastic/es-distributed (Team:Distributed)

idegtiarenko commented on server/src/main/java/org/elasticsearch/snapshots/SnapshotShardsService.java:
```diff
@@ -248,9 +248,8 @@ private void startNewSnapshots(List<SnapshotsInProgress.Entry> snapshotsInProgre
     }
 
-    private void startNewShards(SnapshotsInProgress.Entry entry, Map<ShardId, IndexShardSnapshotStatus> startedShards) {
+    private void startNewShardSnapshots(SnapshotsInProgress.Entry entry, Map<ShardId, IndexShardSnapshotStatus> startedShards) {
```
May be `snapshotNewShards`? I believe we use `snapshot` as a verb quiet often. Also all shards belong to the same snapshot?

pxsalehi replied:
I just tried to keep the names aligned with the existing methods. The caller of this is called `startNewSnapshots` and since each snapshot is broken down into snapshots of the shards (of the indices), I thought `startNewShardSnapshot` might fit better. I kept the `start*` naming since it conveys that the tasks are forked off. Is the naming confusing?

I think the `new` here just means those listed in the snapshotsInProgress, and I think, yes, all shards called here are within the same snapshot.

henningandersen commented:
I wonder if this approach is necessarily better. IIUC, we currently keep the one thread looping until the end. In the new approach, we risk the first (or an early) shard filling up the snapshot queue before the last tasks are added to the queue. This could mean that shards with no changes being scheduled after large file uploads?

original-brownbear commented:
> This could mean that shards with no changes being scheduled after large file uploads?

We figured this was somewhat unlikely (to happen a lot) and outweigh by the fact that we now multi-thread the starting of shard snapshots (which is the main motivation for this change, as we realized that running this single-threaded could take a long time on large data nodes).
But I guess if we want to be stricter about this, we could implement this differently by putting all the shards to start in a queue and then running threads (one of which would be the original thread that we fork from) that poll from that queue. That way we get even better ordering at the price of a little more code? Think it's worth it?

pxsalehi commented:
(Apologies for using too many vague/relative terms like many/few/large! But I need to somehow ask my question!)

For me one missing piece to evaluate the worst case impact of a file upload on unchanged shard snapshots, is the following: could the size of single files (to be uploaded) be extremely large? What would be a realistic "average" case? Or is it hard to tell since it could be wildly different depending on the settings/load? It seemed to me that many not-huge files to uploads is the common case. (Also another assumption from our talk, it seems often what happens during a snapshot is that most shards have no changes, and only few of them need to upload files.)

henningandersen commented:
We generally try to not fill the queue with all tasks, for instance in `BlobStoreRepository.snapshotShard` which only schedules as many threads as there are. I assume to allow other work to interleave?

If we do all shards in parallel, we risk getting `#shards * thread_pool_size` into the queue. It seems like the "throttling" we do per shard is no longer useful then? This could mean a duration of many minutes if not hours of no response on the `SNAPSHOT` thread pool. I think that could block restores, new followers and single-index snapshots? Am I missing something essential here?

I think this would be true also if we queue up the shard level execution and only do n jobs in parallel.

original-brownbear commented:
> Am I missing something essential here?

You're not directly missing anything actually. But I think there is a tradeoff here. This change makes it so that running a snapshot on a 10k shard warm-node may block the snapshot pool for ~15 min (assuming ~0.5s per shard snapshot and 5 snapshot threads max) vs. taking 75min to run a single snapshot without blocking the pool.
Taking the 15 min over the 75 min and accepting the blocking seems like the right tradeoff to me, especially when it's this easy to implement?
Also, note that we are doing the delete side where we're not being clever about doing any sort of hand-crafted work stealing queue like we do for file uploads either and no complaints about it have come in ever as far as I'm aware :)

> I think this would be true also if we queue up the shard level execution and only do n jobs in parallel.

True, the advantage of that approach would be that we'd have a harder guarantee on running the metadata work right before the file uploads (and it's a little faster potentially I guess).

pxsalehi commented:
I'm trying to understand how much of this discussion relates to the original issue (https://github.com/elastic/elasticsearch/issues/83408). 

If blocking the snapshot threadpool during a snapshot is a concern, then what would be an acceptable approach that addresses the issue and avoids blocking the threadpool? One thread going through the shard snapshot tasks, calculating which snapshots do not upload a file, doing all of those in the same thread, and then again in the same thread going through the shard snapshots that do have a change, and only fork-off the file-uploads? 

Basically, everything else other than file upload should be done on the thread going through the shards, and actually in two passes? No snapshotting of unchanged shards in parallel? Considering having many thousand unchanged shard snapshots and few shards actually uploading files is a normal occurrence, as Armin mentioned, parallelizing snapshotting of unchanged shards seems to reduce snapshot time, which I guess is an improvement.

My question is that if we do not want to parallelize shard snapshot tasks, then would the approach I mentioned above have any real improvement for the issue? Currently, all unchanged snapshots happen in the same thread, which doesn't seem that different than the two phase approach!

henningandersen commented:
Armin and I had a brief conversation on this and while the existing mechanism for limiting the number of upload threads is somewhat broken, running the outer level tasks in parallel will slightly increase this brokennes. We prefer to instead fix this for good by maintaining enough data structures to just fill the snapshot pool with the right amount of work without exhausting it for potentially hours.

Something like 2 queues (the other level work and the actual file upload work) and a counter for how many jobs are active should do. When any job finishes it checks the outer level queue first and once that is depleted it does the actual file upload jobs.

This also ensures that all no-change shards are done prior to any file uploads. I think Armin intends to chat sync with you on this.

pxsalehi committed:
temp

pxsalehi committed:
Revert "temp"

This reverts commit 88847d4183185f8ec4f5fec846ce13bd0cda5a61.

pxsalehi committed:
Use queues and workers to handle shard snapshot work

pxsalehi committed:
cleanup, minor corrections

pxsalehi committed:
remove test logging

pxsalehi committed:
Create only as many workers as needed

pxsalehi commented:
@elasticmachine please run elasticsearch-ci/part-1

(I think, test failure was unrelated to the PR. I opened: https://github.com/elastic/elasticsearch/issues/88615)

pxsalehi committed:
add test, minor renames

pxsalehi committed:
Add some asserts, simplify test

pxsalehi reviewed (commented):
@original-brownbear ~While still a draft PR,~ I have some questions about the implementation. Could you please take a look at what I have so far?

pxsalehi committed:
Edit changelog, remove redundant check

pxsalehi commented:
@elasticmachine please update branch

pxsalehi commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'master' into ps-fork-off-snapshot-calls

DaveCTurner commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -3517,26 +3505,16 @@ public int getReadBufferSizeInBytes() {
-        private final Collection<String> blobsToDelete;
-
-        ShardSnapshotMetaDeleteResult(IndexId indexId, int shardId, ShardGeneration newGeneration, Collection<String> blobsToDelete) {
-            this.indexId = indexId;
-            this.shardId = shardId;
-            this.newGeneration = newGeneration;
-            this.blobsToDelete = blobsToDelete;
-        }
-    }
+    private record ShardSnapshotMetaDeleteResult(
+        IndexId indexId,
+        int shardId,
+        ShardGeneration newGeneration,
+        Collection<String> blobsToDelete
+    ) {}
```
workflow nit: it's _so_ much easier on reviewers to make cosmetic changes like this as a separate PR. I'm definitely +1 on making cosmetic improvements as we find them, but -1 on piling them in to more complex changes like this.

pxsalehi replied:
https://github.com/elastic/elasticsearch/pull/88721

pxsalehi committed:
Move out cosmetic changes

pxsalehi committed:
Merge branch 'ps-fork-off-snapshot-calls' of github.com:pxsalehi/elasticsearch into ps-fork-off-snapshot-calls

pxsalehi committed:
Move out unrelated test change

pxsalehi commented on server/src/test/java/org/elasticsearch/repositories/fs/FsRepositoryTests.java:
```diff
@@ -149,8 +148,8 @@ public void testSnapshotAndRestore() throws IOException, InterruptedException {
             IndexCommit incIndexCommit = Lucene.getIndexCommit(Lucene.readSegmentInfos(store.directory()), store.directory());
             Collection<String> commitFileNames = incIndexCommit.getFileNames();
             final PlainActionFuture<ShardSnapshotResult> future2 = PlainActionFuture.newFuture();
+            IndexShardSnapshotStatus snapshotStatus = IndexShardSnapshotStatus.newInitializing(shardGeneration);
```
Since these changes are also unrelated, I moved them out to https://github.com/elastic/elasticsearch/pull/88723.

pxsalehi committed:
Undo cosmetic changes to SnapshotShardsService

pxsalehi commented:
I have also undone the changes to `SnapshotShardsService` in favour of https://github.com/elastic/elasticsearch/pull/88707.

pxsalehi committed:
Merge remote-tracking branch 'upstream/master' into ps-fork-off-snapshot-calls

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -415,6 +418,12 @@ protected void doStart() {
             throw new IllegalArgumentException("the chunk size cannot be negative: [" + chunkSize + "]");
         }
+        shardSnapshotWorkerPool = new ShardSnapshotWorkerPool(
+            threadPool.info(ThreadPool.Names.SNAPSHOT).getMax(),
```
We want to upload at the maximum parallelism possible so max pool size like you use it here is what we want.

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -2882,46 +2895,20 @@ public void snapshotShard(SnapshotShardContext context) {
                 context.onResponse(shardSnapshotResult);
             }, context::onFailure);
-            if (indexIncrementalFileCount == 0) {
+            if (indexIncrementalFileCount == 0 || filesToSnapshot.isEmpty()) {
```
Makes sense, neat fix saving some redundant tasks :)

pxsalehi replied:
Well, with the current PR this was necessary, as otherwise, we'd be creating a `GroupedActionListener` of size 0.

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -415,6 +418,12 @@ protected void doStart() {
         if (chunkSize != null && chunkSize.getBytes() <= 0) {
             throw new IllegalArgumentException("the chunk size cannot be negative: [" + chunkSize + "]");
         }
+        shardSnapshotWorkerPool = new ShardSnapshotWorkerPool(
```
++ makes sense

pxsalehi commented on server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkerPool.java:
```diff
@@ -0,0 +1,137 @@
+
+    private void startWorker(final String workerId) {
+        logger.debug("starting snapshot pool worker {}", workerId);
+        executor.execute(() -> {
```
Thank you both for the feedback.

> I think the answer here is "mostly no".

That's the part I am not entirely sure about when using two queues. I don't like the fact that it is not entirely predictable! It would be great to make sure there is a certain order regarding the sub-tasks, that will always be maintained when snapshotting. I am just not sure if it's an overkill, and don't want to add unnecessary code.

If it is not too much unnecessary code, I could use the `startTime` that is recorded in the `SnapshotsInProgress.Entry`. I think this is set once on master when the snapshot request gets processed, so it seems to stay unchanged afterwards. I could use a priority queue that sorts by snapshot `startTime` -> `snapshotId` -> task type(metadata work or file upload). I'm adding `startTime` to the mix, since ordering by `snapshotId` might still result in some weird corner case, where some snapshot keeps getting pushed back, due to ordering by  names/UUIDs. 

Would there be a downside to adding `startTime` to the ordering?

DaveCTurner replied:
> this only matters if snapshots are triggered at a rate that's higher than the rate at which they can be completed.

I don't think that's quite true - there can temporarily be a bit spike in snapshot work to do even if in the long run we will be able to get through it all. For instance if a hot node fails and a bunch of replicas get promoted then that's going to cause a one-off bigger snapshot job. By prioritising this work over the work for later-starting ones we would typically finish the big snapshot sooner. Snapshots that finish sooner are kind of more useful because of how that reduces the window of uncertainty about which data is/isn't included.

DaveCTurner replied:
> To me it seems preferable to just always do the meta work first, then the file uploads.

I think I agree here, but that means we should also check for new metadata work between file uploads right?

pxsalehi committed:
Use a priority queue for both tasks

pxsalehi committed:
Adapt tests

pxsalehi committed:
Add test

pxsalehi committed:
Simplify mock

pxsalehi committed:
Merge remote-tracking branch 'upstream/main' into ps-fork-off-snapshot-calls

pxsalehi committed:
Merge remote-tracking branch 'upstream/main' into ps-fork-off-snapshot-calls

pxsalehi committed:
Minor cleanup and rename

pxsalehi committed:
One more assert

pxsalehi commented on server/src/test/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkersTests.java:
```diff
@@ -0,0 +1,208 @@
+        assertThat(repo.finishedShardSnapshots(), equalTo(enqueuedSnapshots));
+    }
+
+    public void testCompareToShardSnapshotTask() {
```
To test that ShardSnapshotTask has priority over file snapshots, I tried to add some tests beyond just this compare test, but it turned out quite tricky to verify that. If you have any suggestions, please let me know.

pxsalehi reviewed (commented):
@henningandersen @DaveCTurner @original-brownbear 
This is now ready for review. Please have another look. After our discussion [here](https://github.com/elastic/elasticsearch/pull/88209#discussion_r925485272), I've decided to prioritize "meta" work over file uploads only within each snapshot, not across all snapshots. IMO, this would avoid parallel snapshots taking longer, and within each snapshot it would still finish unchanged shards first.

pxsalehi committed:
Always increment finishedShardSnapshotTasks

pxsalehi commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into ps-fork-off-snapshot-calls

original-brownbear reviewed (commented):
I left a couple of questions + comments mostly on the mechanical details of things. Overall this looks nice and exactly as we want it IMO :) So mainly just the sort order left to discuss here I think.

pxsalehi committed:
Merge remote-tracking branch 'upstream/main' into ps-fork-off-snapshot-calls

pxsalehi committed:
Package private size()

pxsalehi committed:
Non-volatile worker count

pxsalehi committed:
Abstract class for snapshot tasks

pxsalehi committed:
Add to queue outside mutex

pxsalehi committed:
Simpler ensureEnoughWorkers

henningandersen commented on server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkers.java:
```diff
@@ -0,0 +1,179 @@
+    private void ensureEnoughWorkers() {
+        assert Thread.holdsLock(mutex);
+
+        int workersToCreate = Math.min(maxWorkers - workerCount, snapshotTasks.size());
```
I think we would max add one worker here ever? Also, using `snapshotTasks.size()` seems wrong, since the task is polled by a worker, but that worker is not free to do other work until it completes it's task (and thus we may still want a new worker even if there is only 1 task in the queue).

Also, as pointed out by Armin, we poll outside the mutex anyway.

I wonder if we could instead just start a worker every time as long as we have budget for it?

original-brownbear replied:
> I wonder if we could instead just start a worker every time as long as we have budget for it?

++ there isn't much value in being tricky here, just starting up to pool-size workers seems like a good idea and removes the rather hard to reason about `size()` call.

pxsalehi replied:
Hmm. I don't really see that happening as of https://github.com/pxsalehi/elasticsearch/blob/5397f5207da55b41b33afb10046d3079833829a1/server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkers.java. Maybe I'm missing something. In your example, since `maxWorker` is 2 and one is busy, when the second task gets queued we'd call `ensureEnoughWorkers` and we'd end up with `min(2-1, 1)` workers, so a new one will be started. (I think [this test](https://github.com/pxsalehi/elasticsearch/blob/3649729263f79e3d80d191406008e4a6e22df688/server/src/test/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkersTests.java#L157-L181) kind of covers a similar case).

henningandersen commented on server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkers.java:
```diff
@@ -0,0 +1,179 @@
+    }
+
+    private void ensureEnoughWorkers() {
+        assert Thread.holdsLock(mutex);
```
I wonder if this method would be simpler to reason about if it was lock-free, i.e., using an `AtomicInteger` for the worker count and just spawn one new thread if there is budget.

henningandersen commented on server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkers.java:
```diff
@@ -0,0 +1,179 @@
+        }
+
+        @Override
+        public int compareTo(ShardSnapshotTask other) {
```
I think we should mark this method `final` since overriding is likely bug-prone.

pxsalehi replied:
Done!

henningandersen commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -376,6 +377,8 @@ public abstract class BlobStoreRepository extends AbstractLifecycleComponent imp
      */
     private final int maxSnapshotCount;
 
+    private ShardSnapshotWorkers shardSnapshotWorkerPool;
```
nit: I'd prefer to name this identical to the type:

```suggestion
    private ShardSnapshotWorkers shardSnapshotWorkers;
```

henningandersen commented on server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkers.java:
```diff
@@ -0,0 +1,179 @@
+        logger.debug("starting snapshot pool worker {}", workerId);
+        executor.execute(() -> {
+            try {
+                while (true) {
```
I wonder if this should instead respawn to the back of the queue when there is more work, allowing other work to use the snapshot thread pool.

I suggest something like:

`startWorker` takes the `ShardSnapshotTask` as argument and executes it. It checks the queue afterwards and respawns (`executor.execute`) if there is more work to do - with the new task as argument.

We only queue excess tasks. I.e., if there is budget in `workerCount`, we always call `startWorker`, otherwise we queue it. We recheck `workerCount` after queueing to catch the edge case of the last worker exiting - and spawns a new task in that case (polling out the added task from the queue).

pxsalehi replied:
@henningandersen Thanks for the clarification. TBH I think we misunderstood [your original suggestion](https://github.com/elastic/elasticsearch/pull/88209#issuecomment-1177696199)! What you are suggesting here makes way more sense that what I have in the PR! :-D  [I was wondering](https://github.com/elastic/elasticsearch/pull/88209#discussion_r925487903) whether we'd need to at least not use the same number of workers as the threadpool size since the workers would keep the threadpool busy as long as there are tasks in the queue!

Putting the lock-free part aside, I've tried out a draft of your suggestion, please see `ShardSnapshotWorkers.java` again. I've done some renaming as well, I guess the name worker doesn't have much meaning anymore. It's more about controlling max concurrent snapshot tasks and the spawning order. Is that correct?

Please let me know if this is in the right direction. I'll keep addressing the rest of the comments.

> We only queue excess tasks. I.e., if there is budget in workerCount, we always call startWorker, otherwise we queue it. We recheck workerCount after queueing to catch the edge case of the last worker exiting - and spawns a new task in that case (polling out the added task from the queue).

Not sure I completely understand this part. We'd have to always queue first, so that we run the tasks in the priority we want.

henningandersen replied:
> I think we misunderstood https://github.com/elastic/elasticsearch/pull/88209#issuecomment-1177696199

I think you got it just right, I think I educated my thinking here based on your work.

> Not sure I completely understand this part. We'd have to always queue first, so that we run the tasks in the priority we want.

I do not think this is important. It will be right most of the time if not all the time (depending on details in the code). Let me take a look at the change you did.

henningandersen replied:
I think the version you have could work out.

Anyway, I wanted to illustrate the lock-free approach. I think this preserves order just as well as the mutex one. Notice that if there is free capacity we always execute. We only queue if there is not. Meaning the queue is empty if there is free capacity. This is not entirely true in the mutex variant (but could be modified there too).

I think I prefer the lock-free version, but I'd also be ok with the mutex version (have comments for it though).

```
public class ShardSnapshotWorkers {
    private static final Logger logger = LogManager.getLogger(ShardSnapshotWorkers.class);

    private final int maxRunningTasks;
    private final Object mutex = new Object();
    private AtomicInteger runningTasksCount = new AtomicInteger();
    private final BlockingQueue<SnapshotTask> snapshotTasks = new PriorityBlockingQueue<>();
    private final Executor executor;
    private final Consumer<SnapshotShardContext> shardSnapshotter;
    private final CheckedBiConsumer<SnapshotShardContext, FileInfo, IOException> fileSnapshotter;

    abstract class SnapshotTask implements Comparable<SnapshotTask>, Runnable {
        protected final SnapshotShardContext context;

        SnapshotTask(SnapshotShardContext context) {
            this.context = context;
        }

        public abstract short priority();

        public SnapshotShardContext context() {
            return context;
        }

        @Override
        public final int compareTo(SnapshotTask other) {
            int res = context.snapshotId().compareTo(other.context.snapshotId());
            if (res != 0) {
                return res;
            }
            return Integer.compare(priority(), other.priority());
        }
    }

    class ShardSnapshotTask extends SnapshotTask {
        ShardSnapshotTask(SnapshotShardContext context) {
            super(context);
        }

        @Override
        public void run() {
            shardSnapshotter.accept(context);
        }

        @Override
        public short priority() {
            return 1;
        }

    }

    class FileSnapshotTask extends SnapshotTask {
        private final FileInfo fileInfo;
        private final ActionListener<Void> fileUploadListener;

        FileSnapshotTask(SnapshotShardContext context, FileInfo fileInfo, ActionListener<Void> fileUploadListener) {
            super(context);
            this.fileInfo = fileInfo;
            this.fileUploadListener = fileUploadListener;
        }

        @Override
        public void run() {
            ActionRunnable.run(fileUploadListener, () -> fileSnapshotter.accept(context, fileInfo)).run();
        }

        @Override
        public short priority() {
            return 2;
        }
    }

    public ShardSnapshotWorkers(
        final int maxRunningTasks,
        final Executor executor,
        final Consumer<SnapshotShardContext> shardSnapshotter,
        final CheckedBiConsumer<SnapshotShardContext, FileInfo, IOException> fileSnapshotter
    ) {
        assert maxRunningTasks > 0;

        logger.info("starting shard snapshot worker pool of max size {}", maxRunningTasks);
        this.maxRunningTasks = maxRunningTasks;
        this.executor = executor;
        this.shardSnapshotter = shardSnapshotter;
        this.fileSnapshotter = fileSnapshotter;
    }

    public void enqueueShardSnapshot(final SnapshotShardContext context) {
        logger.trace("enqueuing shard snapshot task [snapshotID={}, indexID={}]", context.snapshotId(), context.indexId());
        runOrQueueTask(new ShardSnapshotTask(context));
    }

    public void enqueueFileSnapshot(
        final SnapshotShardContext context,
        final FileInfo fileInfo,
        final ActionListener<Void> fileUploadListener
    ) {
        logger.trace(
            "enqueuing shard snapshot file upload task [snapshotID={}, indexID={}, file={}]",
            context.snapshotId(),
            context.indexId(),
            fileInfo.name()
        );
        FileSnapshotTask task = new FileSnapshotTask(context, fileInfo, fileUploadListener);
        runOrQueueTask(task);
    }

    private void runOrQueueTask(SnapshotTask task) {
        if (incrementRunningTasks()) {
            spawnTask(task);
        } else {
            snapshotTasks.add(task);
            if (incrementRunningTasks()) {
                pollAndSpawn();
            }
        }
    }

    private void pollAndSpawn() {
        SnapshotTask polledTask = snapshotTasks.poll();
        if (polledTask != null) {
            spawnTask(polledTask);
        } else {
            int decremented = runningTasksCount.decrementAndGet();
            assert decremented >= 0;
        }
    }

    private void spawnTask(SnapshotTask task) {
        executor.execute(() -> runOneTask(task));
    }

    private boolean incrementRunningTasks() {
        return runningTasksCount.getAndUpdate(v -> v < maxRunningTasks ? v + 1 : v) < maxRunningTasks;
    }

    // for testing
    int size() {
        synchronized (mutex) {
            return runningTasksCount.get();
        }
    }

    private void runOneTask(final SnapshotTask task) {
        try {
            logger.trace("running snapshot task {}", task); // TODO: toString()?
            task.run();
        } finally {
            pollAndSpawn();
        }
    }

}
```

henningandersen reviewed (commented):
One more comment. I think there could be a liveness issue with current implementation (though it might go away with more mutex or lock-free). I think this suggestion fixes it and is more fair to other ongoing work.

pxsalehi committed:
Create workers in the constructor

pxsalehi commented on server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkers.java:
```diff
@@ -0,0 +1,179 @@
+
+        @Override
+        public int compareTo(ShardSnapshotTask other) {
+            int res = context.snapshotId().compareTo(other.context.snapshotId());
```
There were no more replies to my question/suggestion [here](https://github.com/elastic/elasticsearch/pull/88209#discussion_r927741572), so I went for this as a first shot. Changing it would be easy once we agree on it. (I would need a bit more code since startTime is not part of the context).

So it should be ALL meta work (across all snapshots of the same repository) first, and then among file snapshots, sort by start time? No need to sort by snapshot name/id? Would start time be a replacement for sorting by snapshot name to prevent concurrent snapshots getting interleaved?

My understanding is that it is desirable to have predictable ordering here, and also finish shard snapshots as quick as possible. If a shard snapshot is unchanged, prioratizing it, and when it comes to file snapshots, finish the relevant ones (for the same snapshot) first, so that snapshots are marked as finished one at a time and do not get interleaved.

pxsalehi committed:
Minor renames

pxsalehi committed:
Merge remote-tracking branch 'upstream/main' into ps-fork-off-snapshot-calls

pxsalehi committed:
Do not use looping workers

pxsalehi committed:
Use threadpool instead of mock in EncryptedRepositoryTests

pxsalehi committed:
Undo changes in RepositoriesServiceTests

pxsalehi committed:
Use threadpool instead of mock in RepositoriesServiceTests

pxsalehi committed:
temp

pxsalehi committed:
Revert "Use threadpool instead of mock in RepositoriesServiceTests"

This reverts commit cd70f35081224211a4f38b2eab001ad81b27ecef.

pxsalehi committed:
Revert "Use threadpool instead of mock in EncryptedRepositoryTests"

This reverts commit a33aaeb19d3e799a600702bc125d92f5ee1e42d5.

pxsalehi committed:
add an executor to threadpool mock

pxsalehi committed:
Revert "temp"

This reverts commit b7394e74784212ffa7e9b179bfb637ca58d1471c.

pxsalehi committed:
no executor, just info

pxsalehi committed:
Merge remote-tracking branch 'upstream/main' into ps-fork-off-snapshot-calls

pxsalehi committed:
Use lock-free instead of mutex

pxsalehi committed:
Use startTime to order tasks

pxsalehi committed:
Cleanup

pxsalehi committed:
More cleanup

pxsalehi commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -415,6 +418,12 @@ protected void doStart() {
         if (chunkSize != null && chunkSize.getBytes() <= 0) {
             throw new IllegalArgumentException("the chunk size cannot be negative: [" + chunkSize + "]");
         }
+        shardSnapshotWorkerPool = new ShardSnapshotWorkers(
```
done in 7ce5c419373effbd3172f6a088c7f83d7098533a

pxsalehi commented on server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkers.java:
```diff
@@ -0,0 +1,179 @@
+            fileInfo.name()
+        );
+        synchronized (mutex) {
+            snapshotTasks.add(new FileSnapshotTask(context, fileInfo, fileUploadListener));
```
Outdated

pxsalehi commented on server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkers.java:
```diff
@@ -0,0 +1,179 @@
+        }
+    }
+
+    public int size() {
```
done

pxsalehi commented on server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkers.java:
```diff
@@ -0,0 +1,179 @@
+            return context;
+        }
+
+        public void run() {
```
Yeah. That sounds better. Done in https://github.com/elastic/elasticsearch/pull/88209/files#diff-a7a97862ca5689e6d801f92ba8b1255b68bc4464ede9755efbfb65685752a966

pxsalehi commented on server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkers.java:
```diff
@@ -0,0 +1,179 @@
+        if (workersToCreate > 0) {
+            logger.debug("starting {} shard snapshot workers", workersToCreate);
+        }
+        while (workersToCreate > 0) {
```
outdated

pxsalehi commented on server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkers.java:
```diff
@@ -0,0 +1,179 @@
+    private final Executor executor;
+    private final Consumer<SnapshotShardContext> shardSnapshotter;
+    private final CheckedBiConsumer<SnapshotShardContext, FileInfo, IOException> fileSnapshotter;
+    private volatile int workerCount = 0;
```
outdated

pxsalehi commented:
@henningandersen @original-brownbear Thank you for the reviews. Please have another look. I've addressed all comments.

I haven't added that case in Henning's code sample where we spawn directly if workers are available. IMO, it is an optimization for a very specific corner case which I don't know if we need. Always queueing and polling seems simpler (unless it has some disadvantage I'm not aware of).

pxsalehi commented on docs/changelog/88209.yaml:
```diff
@@ -0,0 +1,6 @@
+pr: 88209
+summary: Parallelize unchanged shard snapshots and limit number of parallel file snapshots
```
I guess this is not clear enough now. I'll change it.

pxsalehi replied:
Done!

pxsalehi committed:
Merge remote-tracking branch 'upstream/main' into ps-fork-off-snapshot-calls

pxsalehi committed:
Update changelog

henningandersen commented on test/framework/src/main/java/org/elasticsearch/index/shard/IndexShardTestCase.java:
```diff
@@ -1010,6 +1010,7 @@ protected ShardGeneration snapshotShard(final IndexShard shard, final Snapshot s
                     snapshotStatus,
                     Version.CURRENT,
                     Collections.emptyMap(),
+                    threadPool.absoluteTimeInMillis(),
```
Also randomize here instead? I would think we do not rely on the specific time here.

pxsalehi replied:
I don't think the test relies on the time. I've changed it to random, here and also all the following ones.

henningandersen commented on x-pack/plugin/core/src/test/java/org/elasticsearch/snapshots/sourceonly/SourceOnlySnapshotShardTests.java:
```diff
@@ -134,6 +134,7 @@ public void testSourceIncomplete() throws IOException {
                         indexShardSnapshotStatus,
                         Version.CURRENT,
                         Collections.emptyMap(),
+                        threadPool.absoluteTimeInMillis(),
```
Randomize instead?

henningandersen commented on x-pack/plugin/core/src/test/java/org/elasticsearch/snapshots/sourceonly/SourceOnlySnapshotShardTests.java:
```diff
@@ -207,6 +209,7 @@ public void testIncrementalSnapshot() throws IOException {
                         indexShardSnapshotStatus,
                         Version.CURRENT,
                         Collections.emptyMap(),
+                        threadPool.absoluteTimeInMillis(),
```
Randomize instead?

henningandersen commented on x-pack/plugin/core/src/test/java/org/elasticsearch/snapshots/sourceonly/SourceOnlySnapshotShardTests.java:
```diff
@@ -238,6 +241,7 @@ public void testIncrementalSnapshot() throws IOException {
                         indexShardSnapshotStatus,
                         Version.CURRENT,
                         Collections.emptyMap(),
+                        threadPool.absoluteTimeInMillis(),
```
Randomize instead?

henningandersen commented on x-pack/plugin/core/src/test/java/org/elasticsearch/snapshots/sourceonly/SourceOnlySnapshotShardTests.java:
```diff
@@ -299,6 +303,7 @@ public void testRestoreMinmal() throws IOException {
                         indexShardSnapshotStatus,
                         Version.CURRENT,
                         Collections.emptyMap(),
+                        threadPool.absoluteTimeInMillis(),
```
Randomize instead?

henningandersen commented on x-pack/plugin/searchable-snapshots/src/test/java/org/elasticsearch/xpack/searchablesnapshots/store/SearchableSnapshotDirectoryTests.java:
```diff
@@ -632,6 +632,7 @@ protected void assertSnapshotOrGenericThread() {
                             snapshotStatus,
                             Version.CURRENT,
                             emptyMap(),
+                            threadPool.absoluteTimeInMillis(),
```
Randomize instead?

henningandersen reviewed (commented):
This direction looks good.

I'd like to see a test case that stresses the multi-threading aspects of the `ShardSnapshotWorkers` class. I have wondered if we could refactor it into something capable of running a fixed number of tasks on a thread-pool, i.e., not know about snapshots. I think that could be useful elsewhere. I mention it mostly in case that makes the multi-threading test easier, that can also be refactored as a follow-up for when we need a similar functionality next time.

pxsalehi committed:
use random time in tests

pxsalehi committed:
Separate TaskRunner, address review

pxsalehi committed:
Merge remote-tracking branch 'upstream/main' into ps-fork-off-snapshot-calls

pxsalehi commented:
@elasticmachine test elasticsearch-ci/part-1
(Test failure doesn't seem to be related to the PR: https://github.com/elastic/elasticsearch/issues/89623)

pxsalehi commented:
@elasticmachine please test elasticsearch-ci/part-1

pxsalehi commented:
@elasticmachine run elasticsearch-ci/part-1

pxsalehi committed:
Merge remote-tracking branch 'upstream/main' into ps-fork-off-snapshot-calls

pxsalehi commented on server/src/test/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkersTests.java:
```diff
@@ -0,0 +1,213 @@
+public class ShardSnapshotWorkersTests extends ESTestCase {
+
+    private static ThreadPool threadPool;
+    private static Executor executor;
```
done

pxsalehi commented on server/src/main/java/org/elasticsearch/repositories/blobstore/BlobStoreRepository.java:
```diff
@@ -3097,10 +3084,10 @@ void ensureNotClosing(final Store store) throws AlreadyClosedException {
     private static ActionListener<Void> fileQueueListener(
         BlockingQueue<BlobStoreIndexShardSnapshot.FileInfo> files,
-        int workers,
+        int groupSize,
```
done

pxsalehi commented on server/src/test/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkersTests.java:
```diff
@@ -0,0 +1,213 @@
+        assertThat(workers.size(), equalTo(maxSize));
+        snapshotBlocker.countDown();
+        // Eventually all workers exit
+        assertBusy(() -> assertThat(workers.size(), equalTo(0)));
```
I've done this for `ThrottledTaskRunnerTests` since the code is a bit differently structured now.

pxsalehi commented on server/src/test/java/org/elasticsearch/repositories/blobstore/ShardSnapshotWorkersTests.java:
```diff
@@ -0,0 +1,213 @@
+        for (int i = 0; i < enqueuedSnapshots; i++) {
+            workers.enqueueShardSnapshot(dummyContext());
+        }
+        assertBusy(() -> assertThat(workers.size(), equalTo(maxSize - 1)));
```
Done

pxsalehi commented on server/src/main/java/org/elasticsearch/common/util/concurrent/ThrottledTaskRunner.java:
```diff
@@ -0,0 +1,81 @@
+        return preUpdateValue < maxRunningTasks;
+    }
+
+    public int runningTasks() {
```
Technically, this would return the number of the running task runners! since whenever a task is finished, we call `pollAndSpawn` again, we could have a task runner that just exists since it finds no tasks to run. But I don't think this would worth adding another counter. Do we care about that distinction?

pxsalehi replied:
done a4ef329d46fb4383a181bb56434febdfab2a33ec

pxsalehi reviewed (commented):
@henningandersen thanks for the suggestions. This is ready for review now. As suggested, I have refactored some parts to a more generic `ThrottledTaskRunner`, added more tests for the multi-threading aspects, and covered the corner case you mentioned for the lock-free approach.

pxsalehi committed:
minor comment changes

henningandersen commented on server/src/test/java/org/elasticsearch/repositories/blobstore/ShardSnapshotTaskRunnerTests.java:
```diff
@@ -0,0 +1,195 @@
+        // Eventually all snapshots are finished
+        assertBusy(() -> {
+            assertThat(repo.finishedShardSnapshots(), equalTo(enqueuedSnapshots));
+            assertThat(taskRunner.runningTasks(), equalTo(0));
+        });
+        assertThat(taskRunner.queueSize(), equalTo(0));
```
I wonder if there is a risk of the `assertBusy` exiting too early, since `runningTasks` could in theory be 0 with many more queued items. Suppose all your runners has just done the decrement and now wants to do the increment. At that time, the `runningTasks()` could be 0?

You will likely have to include either the queue size or the file-snapshot task check in the assert busy. If you include queue-size it must be checked before the `runningTasks` check.

pxsalehi replied:
Maybe there is a misunderstanding but `finishedShardSnapshots` means all file/meta tasks are finished, not just the meta work. I think the assertBusy there waits until all tasks (of both type) are finished.

henningandersen commented on server/src/test/java/org/elasticsearch/common/util/concurrent/ThrottledTaskRunnerTests.java:
```diff
@@ -0,0 +1,123 @@
+
+        @Override
+        public int compareTo(TestTask o) {
+            return randomIntBetween(0, 1000);
```
I'd think this should be between (-1000, 1000).

But I can imagine `PrioritizedBlockingQueue` not working with such a compare. I think it expects objects to adhere to the `compare` contract, which requires stability and transitivity.

Can we assign a random number to the task and compare the tasks using that number instead?

Can we also validate that we execute tasks in order?

pxsalehi replied:
done in 56619c0eb675ec0401f7ddf73b7a1950bf03677c
I added a new test for the ordering in 79331f810c3a4f21152d57ad3d99a0356fdbf5af

henningandersen reviewed (commented):
Thanks for extracting the logic, looks great and will definitely be useful elsewhere.

I left more comments. I'd like validation of the order of execution of tasks in `ThrottledTaskRunner` too.

pxsalehi commented:
@elasticmachine update branch

elasticmachine committed:
Merge branch 'main' into ps-fork-off-snapshot-calls

elasticmachine committed:
Merge branch 'main' into ps-fork-off-snapshot-calls

pxsalehi commented:
I looked into the [previous failures](https://gradle-enterprise.elastic.co/s/lpiy4ql6ecvxc/tests/:server:internalClusterTest/org.elasticsearch.snapshots.CloneSnapshotIT/testRemoveFailedCloneFromCSWithQueuedSnapshotInProgress?top-execution=1), since they were both related to snapshotting. From the previous test failures, it seems `ConcurrentSnapshotsIT.testMasterFailoverOnFinalizationLoop` was failing due to recent changes on main, and is now fixed on main and doesn't fail here anymore. However, `CloneSnapshotIT.testRemoveFailedCloneFromCSWithQueuedSnapshotInProgress` seems to rarely fail on this PR. Not sure why! I can make it time out on a slow system, but can't tell whether the test assumption doesn't match anymore with how we schedule the snapshotting tasks, or there is an issue in the PR. It seems that the snapshot cloning step of the test doesn't return and times out. I'd have to spend some more time on it. 

CC: @original-brownbear maybe we could look into it together.

pxsalehi committed:
Merge remote-tracking branch 'upstream/main' into ps-fork-off-snapshot-calls

pxsalehi committed:
Rename to PrioritizedThrottledTaskRunner

pxsalehi committed:
Comment methods only for testing

pxsalehi committed:
Use a name in the logger

pxsalehi commented on server/src/test/java/org/elasticsearch/repositories/blobstore/ShardSnapshotTaskRunnerTests.java:
```diff
@@ -0,0 +1,195 @@
+
+        public void snapshotShard(SnapshotShardContext context) {
+            int filesToUpload = randomIntBetween(0, 10);
+            if (filesToUpload == 0) {
+                finishedShardSnapshots.incrementAndGet();
+            } else {
```
No, I think it's different. When there is no file to upload, we mark the snapshot as done/finished. The increment at the end of the if/else is for the task, not the snapshot.

henningandersen replied:
Thanks, I missed that.

pxsalehi commented on server/src/test/java/org/elasticsearch/common/util/concurrent/ThrottledTaskRunnerTests.java:
```diff
@@ -0,0 +1,123 @@
+            }));
+            assertThat(taskRunner.runningTasks(), equalTo(maxTasks));
+        }
+        taskBlocker.countDown();
```
Why? as the comment suggests, the point is to make sure we don't go over maxTasks. `newTasks -1` is just a random value between 0 and 9.

henningandersen replied:
Yes, we expect one task to execute and the rest to be queued. I see no reason not to validate that they are in the queue?

Just like the validation at the end that the queue is now empty.

pxsalehi replied:
ok, now I know what you mean! :-) 79331f810c3a4f21152d57ad3d99a0356fdbf5af

pxsalehi committed:
Use generic pool in test

pxsalehi committed:
Rename test

pxsalehi committed:
Use threads not threadpool for test

pxsalehi committed:
Add a priority to the TestTask

pxsalehi committed:
Add more test and assertion

pxsalehi committed:
Sync threads in test

pxsalehi committed:
Merge remote-tracking branch 'upstream/main' into ps-fork-off-snapshot-calls

pxsalehi commented on server/src/main/java/org/elasticsearch/common/util/concurrent/ThrottledTaskRunner.java:
```diff
@@ -0,0 +1,81 @@
+ * ThrottledTaskRunner performs the enqueued tasks in the order dictated by the natural ordering
+ * of the tasks, limiting the max number of concurrently running tasks.
+ */
+public class ThrottledTaskRunner<T extends Comparable<T> & Runnable> {
```
Thanks for the suggestion. Done in 3a4a28eca0d3b264eda25da74390a96b11b7d648

pxsalehi commented on server/src/main/java/org/elasticsearch/common/util/concurrent/ThrottledTaskRunner.java:
```diff
@@ -0,0 +1,81 @@
+        return runningTasks.get();
+    }
+
+    public int queueSize() {
```
doen a4ef329d46fb4383a181bb56434febdfab2a33ec

pxsalehi commented on server/src/test/java/org/elasticsearch/common/util/concurrent/ThrottledTaskRunnerTests.java:
```diff
@@ -0,0 +1,123 @@
+        }
+    }
+
+    public void testThrottledTaskRunner() throws Exception {
```
done 865ceca6e6f0a7021b9db7439071d0fb7c9370cd

pxsalehi commented on server/src/test/java/org/elasticsearch/common/util/concurrent/ThrottledTaskRunnerTests.java:
```diff
@@ -0,0 +1,123 @@
+    }
+
+    public void testEnqueueSpawnsNewTasksUpToMax() throws Exception {
+        int maxTasks = randomIntBetween(1, threadPool.info(ThreadPool.Names.SNAPSHOT).getMax());
```
I changed it in the whole file to GENERIC. I guess I forgot to change it once taking it out of the snapshot context.

pxsalehi commented:
@henningandersen Thanks for the detailed feedback. I addressed all your comments. Please have another look. Meanwhile I'll look into that previous CI failure to see if it is related or I was just paranoid!

pxsalehi commented:
> @henningandersen Thanks for the detailed feedback. I addressed all your comments. Please have another look. Meanwhile I'll look into that previous CI failure to see if it is related or I was just paranoid!

I am not able to reproduce that issue on the branch. For the record, it used to timeout on different asserts in the test, the couple of times that it happened (not just a specific one), and that was on a pretty slow system. I think, it is safe to dismiss that.

henningandersen commented on server/src/test/java/org/elasticsearch/common/util/concurrent/PrioritizedThrottledTaskRunnerTests.java:
```diff
@@ -0,0 +1,177 @@
+
+        @Override
+        public int compareTo(TestTask o) {
+            return Integer.compare(priority, o.priority);
```
Can we return `priority - o.priority` instead? `Integer.compare`normalizes to `-1, 0, 1`, thus we do not test that any other values has the right meaning.

pxsalehi replied:
@henningandersen the `compareTo` is only used by the priority queue, which doesn't care for the actual distance between the two priorities, I think! So I don't understand why does this matter? Could you please elaborate why is this important?

henningandersen replied:
This is part of a test verifying that `PrioritizedThrottledTaskRunner` executes tasks in the right order when presented with *any* comparable object. But the test here is restricted to only verify using a sub-set of such comparable objects that use "normalized" return values in `compareTo`.

Arguing in terms of the implementation seems invalid, the purpose of the tests is to demonstrate that the implementation works under as many circumstances as possible. Given the simple change to demonstrate this under more circumstances here, I think we should make the change.

pxsalehi replied:
I've done the change already. Thanks for the explanation! :)

henningandersen reviewed (commented):
This looks good to me now, but I'd like @original-brownbear to do the final review on this.

pxsalehi committed:
Merge remote-tracking branch 'upstream/main' into ps-fork-off-snapshot-calls

pxsalehi committed:
do not use Integer.compare

original-brownbear commented on server/src/main/java/org/elasticsearch/common/util/concurrent/PrioritizedThrottledTaskRunner.java:
```diff
@@ -0,0 +1,86 @@
+    }
+
+    // visible for testing
+    protected void pollAndSpawn() {
```
Can we add some commentary on how this loop works. In particular, on why we need to peek the queue. This is somewhat hard to follow for me and will be even harder to follow for future readers of this code.

pxsalehi replied:
Good point!

original-brownbear commented on server/src/main/java/org/elasticsearch/repositories/blobstore/ShardSnapshotTaskRunner.java:
```diff
@@ -0,0 +1,140 @@
+            this.context = context;
+        }
+
+        public abstract short priority();
```
Should we just use `int` here. `short` is kind of pointless isn't it, might even be a net negative for comparison performance?

original-brownbear reviewed (commented):
This looks good to me, just one point where I asked for some docs because I'm having trouble understanding the code and one rather trivial detail and this should be good to go :)

pxsalehi committed:
Address review comments

pxsalehi commented:
@original-brownbear @henningandersen All done! Please check again.

original-brownbear commented on server/src/test/java/org/elasticsearch/common/util/concurrent/PrioritizedThrottledTaskRunnerTests.java:
```diff
@@ -0,0 +1,177 @@
+        for (int i = 0; i < enqueued; i++) {
+            new Thread(() -> {
+                try {
+                    threadBlocker.countDown();
```
NIT: could use `CyclicBarrier` here instead which seems to be the correct primitive?

pxsalehi replied:
We could. But since we need only a one-time barrier, the latch is enough I think. I've seen it used like that in several places in the code.

henningandersen replied:
I'd agree with Armin here, `CyclicBarrier` is good for a rendezvous interaction. And saves a line of code too.

But fine to leave as is...

original-brownbear reviewed (approved):
LGTM, I'm good with this one now as is. Performance seems alright too, gave it a quick test run benchmarking. Unless @henningandersen has anything open? I think we're good to go here.

pxsalehi commented:
Thanks Henning and Armin!
